lib/Service/FaceClusterAnalyzer.php

```diff
@@ -100,10 +100,99 @@ public function calculateClusters(string $userId) {
 			foreach ($notYetClustered as $item) {
 				$detection = $item[0];
 				$this->faceDetections->assocWithCluster($detection, $cluster);
-				if ($cluster->getTitle()) {
-					$this->tagManager->assignTags($detection->getFileId(), [$cluster->getTitle()]);
+			}
+		}
+
+		$this->pruneClusters($userId);
+	}
+
+	/**
+	 * @throws \OCP\DB\Exception
+	 */
+	public function pruneClusters(string $userId) {
+		/**
+		 * @var $clusters FaceCluster[]
+		 */
+		$clusters = $this->faceClusters->findByUserId($userId);
+
+		if (count($clusters) === 0) {
+			$this->logger->debug('No face clusters found');
+			return;
+		}
+
+		foreach ($clusters as $cluster) {
+			/**
+			 * @var $detections FaceDetection[]
+			 */
+			$detections = $this->faceDetections->findByClusterId($cluster->getId());
+
+			$filesWithDuplicateFaces = $this->findFilesWithDuplicateFaces($detections);
+			if (count($filesWithDuplicateFaces) === 0) {
+				continue;
+			}
+
+			$centroid = $this->calculateCentroidOfDetections($detections);
```

Comment:
Can you explain what a centroid is and what is the selection logic below?
Are all $detections here of the same face?

How does the code avoids being confused in the case of multiple detection of several faces in the same image? A cluster is specific to a person and so we only work with detections of the same face in all this block, right?

Reply:
> A cluster is specific to a person and so we only work with detections of the same face in all this block, right?

Yes, then we calculate the arithmetic mean (the centroid) of all face vectors in this cluster. Then we group face detections by file id and for all file ids keep only the face detection that is closest to the centroid.

Reply:
Ok so the centroid is computed on all detections and not only the ones on the files with duplicates, so it makes sense to take the one closest to the centroid.